### PR TITLE
Draw callgraph visualisation over polyglot code

### PR DESCRIPTION
**Pull request by snim2**
This PR has been tested with JRuby+JavaScript and JRuby+SimpleLanguage code. Example screenshots are shown below. You may notice that code within the language boxes (i.e. code not written in Ruby) is not visualised. This is because the JS and SL Truffle interpreters do not add the necessary information to the callgraph log.

polyglot_callgraph_js
polyglot_callgraph_sl

### CODE REVIEW

**Comment by ltratt** [comment]
Why are line numbers possibly (always?) inaccurate in polyglot programs? [I ask this naively -- I have no idea of the particular technical limitations in this case.]

**Comment by snim2** [comment]
When polyglot code is exported, Eco has to add some boilerplate to the
original program. This exposes the JS or SL code to JRuby.

Also, between the user pressing 'Profile' and the visualisation appearing
on screen, there is time for the user to edit their code, which may make
the line numbers in the JRuby callgraph invalid.

File: lib/eco/export/jruby.py [hunk 1]
```diff
@@ -381,3 +389,63 @@ def _profile(self):
...
+                            self._annotate_text(callsite_version.callsite.line,
+                                                method.name,
+                                                { method.name : method.is_mega },
+                                                JRubyMorphismLine)
+                        num_calls += 1
+
+    def _annotate_text(self, lineno, text, annotation, klass):
+        """Annotate a node on a given line with a given symbol name.
+        """
+        if lineno < 0:
```
Comment (ltratt) [inline lib/eco/export/jruby.py]: How can `lineno < 0` be true?

File: lib/eco/export/jruby.py [hunk 2]
```diff
@@ -381,3 +389,63 @@ def _profile(self):
...
+            return
+        # Attempt to find 'text' on 'lineno'.
```
Comment (ltratt) [inline lib/eco/export/jruby.py:416]: I'm not saying that there's a nicer approach than this, but I wonder if having so much code in the `try` block is a good plan?

**Comment by ltratt** [comment]
I've made a couple of minor comments and have one over-arching thing. Can we be consistent about using `"` or `'` for strings? I think the Eco codebase uses the former, but I don't care, as long as we're consistent.

**Comment by ltratt** [comment]
[I'm also not quite sure about formatting function arguments with each on its own line, but that's a slightly harder thing to make consistent easily.]

File: lib/eco/export/jruby.py (hunk 1 above)
Comment (snim2) [inline lib/eco/export/jruby.py]: This happens when method is called which has not been defined by the programmer. For example,  in the snippet:

``` ruby
10.times do
    puts "hello"
end
```

`times` is a method on the object `10`, but the user has not defined `times`, so according to the callgraph log, the source section from which `times` is defined starts on line -1. Below is an example of (part of) a callgraph from one of my examples, you can see there are two examples in that digraph of methods defined on line -1:

test_times_callgraph

Comment (ltratt) [inline lib/eco/export/jruby.py]: Aha, OK.

File: lib/eco/export/jruby.py (hunk 2 above)
Comment (snim2) [inline lib/eco/export/jruby.py:416]: It isn't ideal, but I'm trying to avoid searching the whole tree, where possible.

**Comment by snim2** [comment]
OK, quotes are now consistent. Ready to squash?

**Pull request merged**